Move to sdk

labbott opened the pull request:
Move some CLI specific features into the SDK

ahl commented on cli/src/cli_builder.rs:
```diff
@@ -4,20 +4,22 @@
 
+use oxide_api::config::Config;
+use oxide_api::context::Context;
+use oxide_api::ResolveValue;
```
I think this should be here; it's basically a CLI parsing construct

labbott replied:
The resolve gets pulled down into Context -> reqwest client `resolve` field so I think it does need to go in the SDK

ahl commented on oxide.json:
```diff
@@ -9812,6 +9812,60 @@
           "size"
         ]
       },
+      "DiskImport": {
```
so: this file comes untouched from omicron/openapi/nexus.json -- we can't modify it locally. It gets automatically updated.

labbott replied:
I really like the idea of generating the builder for this type. Does it make sense to either 1) Make the same change in nexus.json or 2) add another openAPI specification or generating internal types?

ahl commented on sdk/src/cmd_disk.rs:
```diff
@@ -0,0 +1,506 @@
+    disk: DiskImport,
+}
+
+pub trait ClientDiskCustom {
```
I'd suggest `ClientCompositeExt` and we use it for any "composite" functionality i.e. composed of multiple client API calls (and the docs could indicate that)

ahl commented on sdk/Cargo.toml:
```diff
@@ -21,7 +21,15 @@ serde_json = { workspace = true }
+thiserror = { workspace = true }
+dirs = { workspace = true }
+tokio = { workspace = true }
+indicatif = { workspace = true }
```
I think we'd want to keep all tty output in the CLI. If one were using the SDK in some headless context it would be odd for us to be emitting status to stdout.

ahl commented on sdk/src/cmd_disk.rs:
```diff
@@ -0,0 +1,506 @@
+        Self { client, disk }
+    }
+
+    pub async fn run(&self) -> Result<(), OxideError> {
```
How much do we want this to look like just some other client operation? If the answer is "yeah, mostly like that" (and I happen to think that's the case) I think we'll want this to be `pub async fn send(&self) -> Result<ResponseValue<?>, Error(types::Error)>` i.e. matching the signature of other client operations.

If we're not using `indicatif` directly (see above) this does raise the question of how we would extract status from the consumer. I think that could be some custom aspect of the builder:

```rust
client.disk_import().with_status_handling(|status| { .. }).send().await;
```

labbott replied:
I went back and forth on this. I was struggling with how to map the errors to the existing error type (probably overthinking) but I think you're right that it should look as similar as possible. I'll give it another look.

ahl replied:
It seems possible (likely?) that `progenitor-client` should grow another error case to make this less awkward.

https://docs.rs/progenitor-client/0.4.0/progenitor_client/enum.Error.html

ahl commented on sdk/src/cmd_disk.rs:
```diff
@@ -0,0 +1,506 @@
+            self.unwind_disk_finalize().await?;
+            self.unwind_disk_delete().await?;
+            return Err(OxideError::UploadThreadFailed);
+        }
```
Perhaps something like:

```suggestion
        if let Some(err) = results.iter().filter(Result::is_err).next() {
            // If any of the upload threads returned an error, unwind the disk.
            eprintln!("one of the upload threads failed");
            self.unwind_disk_bulk_write_stop().await?;
            self.unwind_disk_finalize().await?;
            self.unwind_disk_delete().await?;
            return Err(err);
        }
```

ahl reviewed (commented):
good progress. Here are some eventual goals I see--not sure if/which of these might be appropriate for this PR:
- authentication becomes a property of the SDK--let's figure out the right abstractions
- Some of the other composite operations could move here such as the ones in `cmd_instance.rs`

labbott commented:
>     * authentication becomes a property of the SDK--let's figure out the right abstractions
> 
>     * Some of the other composite operations could move here such as the ones in `cmd_instance.rs`

I think both of those are good goals, ideally once we've figured out the correct abstraction it will be easy to convert other commands over.

labbott committed:
Move config and context from the cli to the SDK

Letting the SDK have access to the same functions to grab the
saved host information is handy

ahl commented on cli/src/cli_builder.rs:
```diff
@@ -4,19 +4,21 @@
 
 // Copyright 2023 Oxide Computer Company
```
if you want: update copyrights

ahl commented on sdk/src/lib.rs:
```diff
@@ -15,4 +16,22 @@ mod generated_sdk;
+
 pub use generated_sdk::*;
+
+#[derive(Error, Debug)]
```
doc comment? Maybe

```suggestion
/// Errors for interfaces related to configuration and authentication.
#[derive(Error, Debug)]
```

ahl reviewed (approved):
thank so much for pushing this through!

ahl commented on cli/src/cli_builder.rs:
```diff
@@ -2,21 +2,23 @@
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-// Copyright 2023 Oxide Computer Company
+// Copyright 2023-2024 Oxide Computer Company
```
Now I feel like a real jerk, but I think we just put the most recent year in our copyright notices. 

```suggestion
// Copyright 2024 Oxide Computer Company
```

https://github.com/oxidecomputer/meta/blob/b71cf2cde756ec8012eeb6822bac470b9ed5ba7e/engineering/code.md?plain=1#L5

labbott replied:
I'm blaming my choice to do this on a previous job with over zealous lawyers.  I don't actually care as long as I don't have to sit through a yearly training explaining the importance of copyright year updates.

ahl replied:
Get 4 lawyers in a room; get 5 answers
